docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,170 @@
+Use `FHU::Filesystem::Exists` instead.
+
+#### std::filesystem::canonical
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::path::lexically_normal
+Use `FHU::Filesystem::LexicallyNormal` instead.
+
+#### std::filesystem::create_directory
+#### std::filesystem::create_directories
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::CreateDirectory` and `FHU::Filesystem::CreateDirectories` instead.
+
+#### std::filesystem::path::parent_path
+Use `FHU::Filesystem::ParentPath` instead.
+
+#### std::filesystem::path::filename
+Use `FHU::Filesystem::GetFilename` instead.
+
+#### std::filesystem::copy_file
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
+Use a combination of open and fextl::string APIs instead of fstream.
+
+### `std::string`
+Use `fextl::string` instead.
+
+#### std::to_string
+Use `fextl::fmt::format` instead.
+
+### APIs that FEX doesn't have a replacement for
+Don't use any of these APIs in FEXLoader/FEXInterpreter. Shoutout to
+[this](https://stackoverflow.com/questions/43056338/standard-library-facilities-which-allocate-but-dont-use-an-allocator) StackOverflow post for this
+huge list.
+
+#### `std::any`
+#### `std::function`
+Ensure the std::function doesn't allocate memory.
+#### `std::valarray`
+#### `std::filebuf`
+#### `std::inplace_merge`
+#### `<stdexcept>`
+#### `std::boyer_moore_searcher`
+#### `std::filesystem::path`
+#### `std::filesystem::directory_iterator`
+#### `std::regex`
+#### `std::thread`
+#### `std::async`
+#### `std::packaged_task`
+#### `std::promise`
+#### `<iostream>`
+#### Remember this is not an all-encompassing list! We may find APIs that still allocate memory and need to be avoided!
+
+### Regular memory allocation routines.
+Don't use these directly as they will the glibc allocator.
+
+#### mmap
+Use `FEXCore::Allocator::mmap`
+
+#### munmap
+Use `FEXCore::Allocator::munmap`
+
+#### malloc
+Use `FEXCore::Allocator::malloc`
+
+#### calloc
+Use `FEXCore::Allocator::calloc`
+
+#### memalign 
+Use `FEXCore::Allocator::memalign`
+
+#### valloc
+Use `FEXCore::Allocator::valloc`
+
+#### posix_memalign
+Use `FEXCore::Allocator::posix_memalign`
+
+#### realloc
+Use `FEXCore::Allocator::realloc`
+
+#### free
+Use `FEXCore::Allocator::free`
+
+#### aligned_alloc
+Use `FEXCore::Allocator::aligned_alloc`
+
+#### __libc_malloc
+**Don't use**
+
+#### __libc_calloc
+**Don't use**
+
+#### __libc_memalign
+**Don't use**
+
+#### __libc_valloc
+**Don't use**
+
+#### __posix_memalign
+**Don't use**
+
+#### __malloc_usable_size
+**Don't use**
+
+## FEX thunk considerations
+Now that we understand the problems of FEX memory allocations. How do these problems affect FEX's thunks?
+
+### Don't use FEXCore APIs that allocate memory through FEX's allocator
+It's as simple as that. Thunks need to use the glibc allocator so their memory allocations end up using the host glibc allocator.
```

Comment:
FEXCore APIs are not accessible from within thunks in the first place (unless explicitly wired up), so this is a bit misleading. Describing why it's undesirable for thunks to access those APIs and which specific allocations unconditionally *mustn't* be allocated through them (i.e. pointers that are guest-facing because they are returned from functions or passed to a callback) seems more instructive here.
The following paragraph would become redundant then too, since it's clear we have more leeway with allocations that aren't guest-facing.

Reply:
Modified